### PR TITLE
Tupha/relative image url uwp renderer uwp

### PR DESCRIPTION
**Pull request by tuanh118**
- Added support for `imageBaseUrl` in host config in order to resolve local image sources in `backgroundImage`, image's `url`, and action's `iconUrl`
- Added a new card with different local image sources
- Modified a host config to add `imageBaseUrl` (with necessary update to some expected test files despite same result)

### CODE REVIEW

**Comment by tuanh118** [comment]
@matthidinger is there anything I need to do for .NET tests to accommodate my new test?

**Comment by nesalang** [comment]
you need to add a check at AllPayloadTests.cs
Please refer to Gilles' commit on 4/12

---
In reply to: [387553756](https://github.com/Microsoft/AdaptiveCards/pull/1437#issuecomment-387553756) [](ancestors = 387553756)

File: source/uwp/AdaptiveCards.sln [hunk 1]
```diff
@@ -36,20 +36,20 @@ Global
 		{CF169157-F487-420C-8B9B-CA3ABE2BB209}.Release|x64.Build.0 = Release|x64
 		{CF169157-F487-420C-8B9B-CA3ABE2BB209}.Release|x86.ActiveCfg = Release|Win32
 		{CF169157-F487-420C-8B9B-CA3ABE2BB209}.Release|x86.Build.0 = Release|Win32
-		{B5DECA6E-D9EA-476D-9B7E-34D3B3044CD8}.Debug|ARM.ActiveCfg = Debug|ARM
```
Comment (nesalang) [inline source/uwp/AdaptiveCards.sln]: adding changes to AdaptiveCards.sln always leads to problems. unless it was intended, please revert it.

File: samples/v1.0/Elements/Image.ImageBaseUrl.json [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+  "backgroundImage": "Assets/background.jpg",
+	"body": [
+		{
+			"type": "Image",
+			"url": "Assets/Adaptive10.png"
+		}
+	],
+  "actions": [
+    {
+      "type": "Action.ShowCard",
```
Comment (RebeccaAnne) [inline samples/v1.0/Elements/Image.ImageBaseUrl.json]: >       [](start = 0, length = 6)

The spacing in this file is inconsistent (some tabs some spaces). Please use tabs in the .json files.

File: source/uwp/Renderer/idl/AdaptiveCards.Rendering.Uwp.idl [hunk 3]
```diff
@@ -2061,8 +2070,11 @@ AdaptiveNamespaceStart
         [propget] HRESULT Id([out, retval] HSTRING* value);
         [propput] HRESULT Id([in] HSTRING value);
 
-        [propget] HRESULT IconUrl([out, retval] Windows.Foundation.Uri** value);
-        [propput] HRESULT IconUrl([in] Windows.Foundation.Uri* value);
+        [propget] HRESULT IconUri([out, retval] HSTRING* value);
+        [propput] HRESULT IconUri([in] HSTRING value);
+
+        [propget] HRESULT IsIconUriRelative([out, retval] boolean* value);
+        [propput] HRESULT IsIconUriRelative([in] boolean value);
```
Comment (RebeccaAnne) [inline source/uwp/Renderer/idl/AdaptiveCards.Rendering.Uwp.idl]: I don't think we want a putter for this.

Comment (RebeccaAnne) [reply source/uwp/Renderer/idl/AdaptiveCards.Rendering.Uwp.idl]: (for all isrelative properties)

---
In reply to: [187156141](https://github.com/Microsoft/AdaptiveCards/pull/1437#discussion_r187156141) [](ancestors = 187156141)

File: source/uwp/Renderer/lib/AdaptiveActionElement.cpp [hunk 4]
```diff
@@ -18,16 +18,7 @@ AdaptiveNamespaceStart
...
-        RETURN_IF_FAILED(GetActivationFactory(
-            HStringReference(RuntimeClass_Windows_Foundation_Uri).Get(),
-            &uriActivationFactory));
-
-        std::wstring iconUrl = StringToWstring(sharedModel->GetIconUrl());
-        if (!iconUrl.empty())
-        {
-            RETURN_IF_FAILED(uriActivationFactory->CreateUri(HStringReference(iconUrl.c_str()).Get(), m_iconUrl.GetAddressOf()));
-        }
+        AssignImageFromUrl(sharedModel->GetIconUrl(), m_iconUri.GetAddressOf(), &m_isIconUriRelative);
```
Comment (RebeccaAnne) [inline source/uwp/Renderer/lib/AdaptiveActionElement.cpp]: >AssignImageFromUrl [](start = 8, length = 18)

Check return value

File: source/uwp/Renderer/lib/Util.cpp [hunk 5]
```diff
@@ -741,4 +741,24 @@ std::string WstringToString(const std::wstring& input)
...
+HRESULT AssignImageFromUrl(const std::string& url, HSTRING* uri, boolean* isUriRelative)
+{
+    if (!url.empty())
+    {
+        RETURN_IF_FAILED(WStringToHString(StringToWstring(url), uri));
+
+        // Try treating uri as absolute and activating it
+        // If it fails, consider it relative
+        ComPtr<IUriRuntimeClassFactory> uriActivationFactory;
+        THROW_IF_FAILED(GetActivationFactory(
```
Comment (RebeccaAnne) [inline source/uwp/Renderer/lib/Util.cpp]: >THROW_IF_FAILED [](start = 8, length = 15)

Change to RETURN_IF_FAILED (don't throw and return from the same function)

File: source/uwp/Renderer/idl/AdaptiveCards.Rendering.Uwp.idl [hunk 6]
```diff
@@ -567,8 +567,11 @@ AdaptiveNamespaceStart
         [propget] HRESULT FallbackText([out, retval] HSTRING* value);
         [propput] HRESULT FallbackText([in] HSTRING value);
 
-        [propget] HRESULT BackgroundImage([out, retval] Windows.Foundation.Uri** value);
-        [propput] HRESULT BackgroundImage([in] Windows.Foundation.Uri* value);
+        [propget] HRESULT BackgroundImageUri([out, retval] HSTRING* value);
+        [propput] HRESULT BackgroundImageUri([in] HSTRING value);
+
+        [propget] HRESULT IsBackgroundImageUriRelative([out, retval] boolean* value);
+        [propput] HRESULT IsBackgroundImageUriRelative([in] boolean value);
```
Comment (RebeccaAnne) [inline source/uwp/Renderer/idl/AdaptiveCards.Rendering.Uwp.idl]: If you haven't already, run this by Matt to make sure he's okay with both the breaking change and the loss of a strong type.

**Comment by tuanh118** [comment]
I talked with @matthidinger today and we decided to get rid of all "is relative" flags and determine that information only in [xaml.cpp](https://github.com/Microsoft/AdaptiveCards/pull/1437/files#diff-fe4934f02789a909f273cce0e859ee3a). This simpler way, compared to the previous approach, actually reduces the work in creating images and avoid exposing the "is relative" flags in the API.

We still need to introduce a breaking change in API, but it's easy for people to change. I have created https://github.com/Microsoft/AdaptiveCards/issues/1439 for the update work

File: source/uwp/Renderer/lib/AdaptiveCard.cpp [hunk 7]
```diff
@@ -303,12 +293,10 @@ AdaptiveNamespaceStart
         adaptiveCard->SetSpeak(HStringToUTF8(m_speak.Get()));
         adaptiveCard->SetLanguage(HStringToUTF8(m_language.Get()));
 
-        if (m_backgroundImage != nullptr)
+        if (m_backgroundImageUri != NULL)
```
Comment (RebeccaAnne) [inline source/uwp/Renderer/lib/AdaptiveCard.cpp]: Now that this is just a string, you can simplify this using a call to the version of HStringToUTF8 that doesn't fail on null values:

adaptiveCard->SetBackgroundImage(HStringToUTF8(m_backgroundImageUri.Get()));

File: source/uwp/Renderer/idl/AdaptiveCards.Rendering.Uwp.idl [hunk 8]
```diff
@@ -1961,6 +1961,9 @@ AdaptiveNamespaceStart
         [propget] HRESULT SupportsInteractivity([out, retval] boolean* value);
         [propput] HRESULT SupportsInteractivity([in] boolean value);
 
+        [propget] HRESULT ImageBaseUrl([out, retval] HSTRING* value);
```
Comment (khouzam) [inline source/uwp/Renderer/idl/AdaptiveCards.Rendering.Uwp.idl]: >ImageBaseUrl [](start = 26, length = 12)

Shouldn't we be consistent in our usage of URI and URL?

File: source/uwp/Renderer/lib/AdaptiveActionElement.cpp [hunk 9]
```diff
@@ -92,14 +84,11 @@ AdaptiveNamespaceStart
     {
         sharedCardElement->SetId(HStringToUTF8(m_id.Get()));
         sharedCardElement->SetTitle(HStringToUTF8(m_title.Get()));
-        
-        if (m_iconUrl != nullptr)
-        {
-            HString urlTemp;
-            m_iconUrl->get_AbsoluteUri(urlTemp.GetAddressOf());
 
+        if (m_iconUri != NULL)
```
Comment (khouzam) [inline source/uwp/Renderer/lib/AdaptiveActionElement.cpp]: >NULL [](start = 25, length = 4)

keep nullptr

File: source/uwp/Renderer/lib/AdaptiveCard.cpp [hunk 10]
```diff
@@ -164,16 +164,7 @@ AdaptiveNamespaceStart
...
-        RETURN_IF_FAILED(GetActivationFactory(
-            HStringReference(RuntimeClass_Windows_Foundation_Uri).Get(),
-            &uriActivationFactory));
-
-        std::wstring imageUri = StringToWstring(sharedAdaptiveCard->GetBackgroundImage());
-        if (!imageUri.empty())
-        {
-            RETURN_IF_FAILED(uriActivationFactory->CreateUri(HStringReference(imageUri.c_str()).Get(), m_backgroundImage.GetAddressOf()));
-        }
+        RETURN_IF_FAILED(WStringToHString(StringToWstring(sharedAdaptiveCard->GetBackgroundImage()), m_backgroundImageUri.GetAddressOf()));
```
Comment (khouzam) [inline source/uwp/Renderer/lib/AdaptiveCard.cpp]: >WStringToHString(StringToWstring( [](start = 25, length = 33)

UTF8ToHString

File: source/uwp/Renderer/lib/AdaptiveCard.cpp (hunk 7 above)
Comment (khouzam) [inline source/uwp/Renderer/lib/AdaptiveCard.cpp]: >NULL [](start = 36, length = 4)

nullptr

File: source/uwp/Renderer/lib/AdaptiveHostConfig.cpp [hunk 11]
```diff
@@ -133,6 +134,18 @@ AdaptiveNamespaceStart
         return S_OK;
     }
 
+    _Use_decl_annotations_
+        HRESULT AdaptiveHostConfig::get_ImageBaseUrl(HSTRING* imageBaseUrl)
```
Comment (khouzam) [inline source/uwp/Renderer/lib/AdaptiveHostConfig.cpp]: >    [](start = 5, length = 3)

Alignment

File: source/uwp/Renderer/lib/AdaptiveImage.cpp [hunk 12]
```diff
@@ -29,16 +29,7 @@ AdaptiveNamespaceStart
...
-        RETURN_IF_FAILED(GetActivationFactory(
-            HStringReference(RuntimeClass_Windows_Foundation_Uri).Get(),
-            &uriActivationFactory));
-        
-        std::wstring imageUri = StringToWstring(sharedImage->GetUrl());
-        if (!imageUri.empty())
-        {
-            RETURN_IF_FAILED(uriActivationFactory->CreateUri(HStringReference(imageUri.c_str()).Get(), m_url.GetAddressOf()));
-        }
+        RETURN_IF_FAILED(WStringToHString(StringToWstring(sharedImage->GetUrl()), m_uri.GetAddressOf()));
```
Comment (khouzam) [inline source/uwp/Renderer/lib/AdaptiveImage.cpp]: >WStringToHString(StringToWstring [](start = 25, length = 32)

UTF8ToHString

File: source/uwp/Renderer/lib/Util.cpp [hunk 13]
```diff
@@ -38,739 +38,745 @@ using namespace std;
...
-    });
-
-    return S_OK;
+		ABI::Windows::Foundation::Collections::IVector<ABI::AdaptiveNamespace::IAdaptiveCardElement *> *items,
+		std::vector<std::shared_ptr<AdaptiveSharedNamespace::BaseCardElement>> &containedElements)
+{
+	containedElements.clear();
+
+	XamlHelpers::IterateOverVector<ABI::AdaptiveNamespace::IAdaptiveCardElement>(items, [&](ABI::AdaptiveNamespace::IAdaptiveCardElement *item) {
+		ABI::AdaptiveNamespace::ElementType elementType;
```
Comment (khouzam) [inline source/uwp/Renderer/lib/Util.cpp]: new line

File: source/uwp/Renderer/lib/XamlBuilder.cpp [hunk 14]
```diff
@@ -394,20 +394,24 @@ AdaptiveNamespaceStart
...
         THROW_IF_FAILED(MakeAndInitialize<AdaptiveImage>(&adaptiveImage));
-        adaptiveImage->put_Url(url);
+        adaptiveImage->put_Uri(*uri);
 
         ComPtr<IAdaptiveCardElement> adaptiveCardElement;
         THROW_IF_FAILED(adaptiveImage.As(&adaptiveCardElement));
         ComPtr<IUIElement> backgroundImage;
         BuildImage(adaptiveCardElement.Get(), renderContext, renderArgs, &backgroundImage);
+        if (backgroundImage == nullptr) {
+            return;
```
Comment (khouzam) [inline source/uwp/Renderer/lib/XamlBuilder.cpp]: newline

File: source/uwp/Renderer/lib/AdaptiveActionElement.cpp [hunk 15]
```diff
@@ -18,16 +18,7 @@ AdaptiveNamespaceStart
...
-        RETURN_IF_FAILED(GetActivationFactory(
-            HStringReference(RuntimeClass_Windows_Foundation_Uri).Get(),
-            &uriActivationFactory));
-
-        std::wstring iconUrl = StringToWstring(sharedModel->GetIconUrl());
-        if (!iconUrl.empty())
-        {
-            RETURN_IF_FAILED(uriActivationFactory->CreateUri(HStringReference(iconUrl.c_str()).Get(), m_iconUrl.GetAddressOf()));
-        }
+        RETURN_IF_FAILED(WStringToHString(StringToWstring(sharedModel->GetIconUrl()), m_iconUri.GetAddressOf()));
```
[the remark this answers is not in the document]
Comment (tuanh118) [reply source/uwp/Renderer/lib/AdaptiveActionElement.cpp]: No. I'll fix it

File: source/uwp/Renderer/idl/AdaptiveCards.Rendering.Uwp.idl (hunk 8 above)
Comment (tuanh118) [reply source/uwp/Renderer/idl/AdaptiveCards.Rendering.Uwp.idl]: I guess URI would be a better (more general) term. Do you think we should change in other places as well? For example, all source names in json payload use URL (`Image`'s url, `Action`'s iconUrl) and shared objects' variables and functions use url as well [Image.h](https://github.com/Microsoft/AdaptiveCards/blob/master/source/shared/cpp/ObjectModel/Image.h#L18)

File: source/uwp/Renderer/lib/Util.cpp [hunk 16]
```diff
@@ -370,104 +367,110 @@ HRESULT GenerateActionProjection(
...
     return S_OK;
-} CATCH_RETURN;
+}
+CATCH_RETURN;
 
 HRESULT GenerateColumnsProjection(
-    const std::vector<std::shared_ptr<AdaptiveSharedNamespace::Column>>& containedElements,
-    ABI::Windows::Foundation::Collections::IVector<ABI::AdaptiveNamespace::IAdaptiveColumn*>* projectedParentContainer) noexcept try
+        const std::vector<std::shared_ptr<AdaptiveSharedNamespace::Column>> &containedElements,
+        ABI::Windows::Foundation::Collections::IVector<ABI::AdaptiveNamespace::IAdaptiveColumn *> *projectedParentContainer) noexcept try
```
Comment (RebeccaAnne) [inline source/uwp/Renderer/lib/Util.cpp]: > *> * [](start = 94, length = 5)

Why are we moving these * over? Most of our other code is consistent with the style before this change.

File: source/uwp/Renderer/lib/Util.cpp [hunk 17]
```diff
@@ -370,104 +367,110 @@ HRESULT GenerateActionProjection(
...
 
     return S_OK;
-} CATCH_RETURN;
+}
+CATCH_RETURN;
 
 HRESULT GenerateColumnsProjection(
-    const std::vector<std::shared_ptr<AdaptiveSharedNamespace::Column>>& containedElements,
-    ABI::Windows::Foundation::Collections::IVector<ABI::AdaptiveNamespace::IAdaptiveColumn*>* projectedParentContainer) noexcept try
+        const std::vector<std::shared_ptr<AdaptiveSharedNamespace::Column>> &containedElements,
```
Comment (RebeccaAnne) [inline source/uwp/Renderer/lib/Util.cpp]: >         [](start = 0, length = 8)

Why this change of style to indent the parameters 8 spaces instead of 4? Most of our code indents these 4 spaces.

File: source/uwp/Renderer/lib/XamlBuilder.cpp [hunk 18]
```diff
@@ -394,20 +394,25 @@ AdaptiveNamespaceStart
     _Use_decl_annotations_
     void XamlBuilder::ApplyBackgroundToRoot(
         ABI::Windows::UI::Xaml::Controls::IPanel* rootPanel,
-        ABI::Windows::Foundation::IUriRuntimeClass* url,
+        HSTRING* uri,
```
Comment (RebeccaAnne) [inline source/uwp/Renderer/lib/XamlBuilder.cpp]: >* [](start = 15, length = 1)

You can just pass the HSTRING here, it doesn't need to be HSTRING*

File: source/uwp/Renderer/lib/Util.cpp (hunk 16 above)
Comment (tuanh118) [reply source/uwp/Renderer/lib/Util.cpp]: I'm not sure why it's saved like that for this file

Comment (RebeccaAnne) [reply source/uwp/Renderer/lib/Util.cpp]: If you don't have any changes you actually want to make to this file, consider checking it out from master to get back to where it started.

---
In reply to: [188026941](https://github.com/Microsoft/AdaptiveCards/pull/1437#discussion_r188026941) [](ancestors = 188026941)

File: source/uwp/Renderer/idl/AdaptiveCards.Rendering.Uwp.idl (hunk 8 above)
Comment (tuanh118) [reply source/uwp/Renderer/idl/AdaptiveCards.Rendering.Uwp.idl]: renamed in 2d6f60b

**Comment by khouzam** [comment]
>             "title": "OK"

No change to that file. Please revert. This should remove all the updated ActivityUpdate test files.

---
Refers to: samples/v1.0/Scenarios/ActivityUpdate.json:111 in 45eb751. [](commit_id = 45eb75197175cd01db8965571a1a907fb04c37f2, deletion_comment = False)
